### PR TITLE
add structure data preprocessing and training 

### PR DESCRIPTION
**Pull request by brandondutra**
This solution is not finished, just splitting the pull request now to start the review process. 

In this pull request, the preprocessing code works via gcloud and the datalab interface for both local and cloud runs. Only local training is currently connected in datalab, and it requires TF 0.12 for now. 

The structured data solution solves classification and regression problems using tf.lear DNN and linear models. A new file, for now called transforms config file, is used to define the transforms. This one transforms file is used in preprocessing and training. I'll give an example a following message.

### CODE REVIEW

**Comment by brandondutra** [comment]
**Example of using cloud preprocessing with datalab magics**

I have two csv data files (gs://cloud-ml-dev_bdt/data*.csv) that contains a regression problem with 3 numeric input columns and 3 categorical input columns. 

The transforms_config_file is located here: gs://cloud-ml-dev_bdt/transforms_config_dnn.json

Note that the config file lists the columns in the csv file, the key columns, the target column, problem type [classification/regression], the model type [dnn/linear], and the transforms [scale, embedding, sparse, etc].

To try cloud preprocessing, 
1) build the package (python setup.py sdist)
2)  run the magic
%ml preprocess --package /content/pydatalab/solutionbox/structured_data/dist/structured_data-0.1.tar.gz --cloud
input_file_path: gs://cloud-ml-dev_bdt/data*.csv
output_dir: gs://BUCKET/preprocessed_output
transforms_config_file: gs://cloud-ml-dev_bdt/transforms_config_dnn.json

Local training only works if tf 0.12 is installed for now.

**Comment by brandondutra** [comment]
**roadmap/todo for future pull requests**
1) get training within datalab finished
2) add prediction code that works off gcloud
3) add the prediction datalab interface
4) revisit the config file: consider splitting it up into a schema/transform file. More thought needs to be made on this.
4) revisit using tf.learn/tf.layers: do not use contrib code

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/task.py [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+
+    # Get the model to train.
+    estimator = get_estimator(output_dir, feature_columns, metadata, 
+                              transform_config, args)
+
+    input_placeholder_for_prediction = get_placeholder_input_fn(metadata, 
+        transform_config)
+
+    # Save the finished model to output_dir/model
+    export_monitor = util.ExportLastModelMonitor(
```
Comment (qimingj) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/task.py]: Does it export the metadata as well? Prediction relies on the metadata in the same dir as the model.

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+  """
+  
+  #TODO(brandondutra): fix to TF 0.12
+  #TODO(brandondutra): allow other flags to be set.
+
+  train_filename = os.path.join(preprocessed_dir, 'features_train*')
+  eval_filename = os.path.join(preprocessed_dir, 'features_eval*')
+  metadata_filename = os.path.join(preprocessed_dir, 'metadata.json')
+  this_folder = os.path.dirname(os.path.abspath(__file__))
+  cmd = ['cd %s &&' % this_folder, 'gcloud alpha ml local train',
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: Side question: could reset session also kill any running child process? That sounds like something useful others would need too.

Are you recommending something like this:
local_preprocess: call code directly 
cloud_preprocess: call code directly 
local_train: call code directly 
cloud_train: use datalab/mlalpha/_cloud_runner.py
local_predict: call code directly
cloud_predict: use datalab/mlalpha/_cloud_predictor.py
local_batch_predict: call code directly
cloud_batch_predict: Does datalab support this? Is the only option gcloud?

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+         '--metadata_path=%s' % metadata_filename,
+         '--output_path=%s' % output_dir,
+         '--transforms_config_file=%s' % transforms_config_file,
+         '--max_steps=%s' % str(max_steps)]
+  if layer_sizes:
+    cmd += ['--layer_sizes %s' % layer_sizes]
+
+  print('Local training, running command: %s' % ' '.join(cmd))
+  _run_cmd(' '.join(cmd))
+
```
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: done

File: solutionbox/structured_data/datalab_solutions/structured_data/trainer/task.py (hunk 1 above)
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/trainer/task.py]: Yes. "additional_assets=[args.metadata_path]" puts a copy of the metadata.json file in the final model dir.

File: solutionbox/structured_data/datalab_solutions/structured_data/_package.py (hunk 2 above)
Comment (brandondutra) [inline solutionbox/structured_data/datalab_solutions/structured_data/_package.py]: Interesting note: in your %%ml magics doc, in the section called 'Model/Job Management' Nikhil made a comment to use gcloud for the job management stuff. If gcloud is going to be a first class concept for datalab, then maybe there is some value in showing the user the gcloud command for training/prediction. But this still does not address your hanging question.

**Pull request merged**